Create JS Samples for Azure On Your Data (#85)
Azure-Samples/azureai-samples: scenarios/bring-your-own-data/bringYourOwnData.md

> # Use Azure OpenAI On Your Data With openai-node Library

Reviewer: ```suggestion
# Use Azure OpenAI On Your Data with openai
```

## Objective

The main objective of this tutorial is to help users understand the process of chatting on top of your own data. By the end of this tutorial, you should be able to:

 - Connect to Azure using AzureOpenAI client
 - Get chat response with your own data

## Time

You should expect to spend 5-10 minutes running this sample. 

## Before you begin

> ### Set Environment Variables

Reviewer: ```suggestion
### Set environment variables
```

To start with let us create a config `.env` file with your project details. This file can be used in this sample or other samples to connect to your workspace. A sample `.env` file is provided below with the variables that you need.

```js

> OPENAI_API_VERSION="Insert the targeted API version here"

Reviewer: ```suggestion
OPENAI_API_VERSION="Insert the desired Azure OpenAI API version here"
```

> AZURE_OPENAI_ENDPOINT="Insert your endpoint here"

Reviewer: ```suggestion
AZURE_OPENAI_ENDPOINT="Insert your Azure OpenAI resource endpoint here"
```

> AZURE_OPENAI_API_KEY="Insert your API Key here"

Reviewer: Let's delete this var and focus on the AAD scenario.

Reviewer: Could we also add the Search env vars?

```

> ### Connect To Your Data Sources

Reviewer: ```suggestion
### Connect to your data source
```

We need to connect to a data source to upload our data. We will set up your own designated data sources using Azure AI Search. 

```js
const dataSources = {
    data_sources: [
        {
            type: "azure_search",
            parameters: {
                endpoint: azureSearchEndpoint,
                index_name: azureSearchIndexName,
                authentication: {
                    type: "api_key",
                    key: azureSearchAdminKey,
                }
            }
        }
    ]
}
```

### Using Azure OpenAI client

> We will access Azure Open AI service through `openai-node` library. 

Reviewer: This part needs to discuss more that we need to import both identity and openai and refer to https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/README.md.

> #### Authenticate Using Microsoft Entra ID

Reviewer: ```suggestion
#### Authenticate using Microsoft Entra ID
```

```js
const { AzureOpenAI } = require("openai");  
const { getBearerTokenProvider, DefaultAzureCredential } = require("@azure/identity");  

const credential = new DefaultAzureCredential();      
const scope = "https://cognitiveservices.azure.com/.default";

const client = new AzureOpenAI({
    azureADTokenProvider: getBearerTokenProvider(credential, scope),
});
```

> ## Run AI Models On Your Data

Reviewer: ```suggestion
## Run AI models on your data
```

Now we can use Azure on your own data with Chat Completions. Providing our search endpoint, key, and index name in the `data_sources` options, any questions posed to the model will now be grounded in our own data. An additional property, `context`, will be provided in the response to show the data the model referenced to answer the question.

```js
async function main() {
    console.log('== Get chat completions on your own data ==');
    const deploymentName = "gpt-4";
    const result = await client.chat.completions.create({
        model: deploymentName, 

>         messages: [{ role: 'user', content: 'Write a poem to celebrate my birthday!' }],

Reviewer: Can we be consistent throughout the document with using single quotes or double quotes?

# Use Azure OpenAI On Your Data with openai

## Objective

The main objective of this tutorial is to help users understand the process of chatting on top of your own data. By the end of this tutorial, you should be able to:

>  - Connect to Azure using AzureOpenAI client

Reviewer 2: ```suggestion
 - Connect to Azure using an Azure OpenAI client
```

 - Get chat response with your own data

## Time

You should expect to spend 5-10 minutes running this sample. 

## Before you begin

### Set environment variables

> To start with let us create a config `.env` file with your project details. This file can be used in this sample or other samples to connect to your workspace. A sample `.env` file is provided below with the variables that you need.

Reviewer 2: ```suggestion
To start, create a config `.env` file with your project details. This file can be used in this sample or other samples to connect to your workspace. A sample `.env` file is provided below with the variables that you need.
```

```js
OPENAI_API_VERSION="Insert the desired Azure OpenAI API version here"
AZURE_OPENAI_ENDPOINT="Insert your Azure OpenAI resource endpoint here"

> AZURE_OPENAI_API_KEY="Insert your API Key here"

Reviewer: ```suggestion
```

AZURE_SEARCH_ENDPOINT="Insert your Search Endpoint here"
AZURE_SEARCH_INDEX="Insert your Search index name here"
AZURE_SEARCH_KEY="Insert your Search API Key here"
```

### Connect to your data source
We need to connect to a data source to upload our data. We will set up your own designated data sources using Azure AI Search and import the values from our `.env` file.

```js
require('dotenv').config();

const azureSearchEndpoint = process.env["AZURE_SEARCH_ENDPOINT"] || "<search endpoint>";
const azureSearchIndexName = process.env["AZURE_SEARCH_INDEX"] || "<search index>";
const azureSearchAdminKey = process.env["AZURE_SEARCH_KEY"] || "<search key>";

const dataSources = {
    data_sources: [
        {
            type: "azure_search",
            parameters: {
                endpoint: azureSearchEndpoint,
                index_name: azureSearchIndexName,
                authentication: {

>                     type: "api_key",

Reviewer 2: Let's replace API key with managed identity. We don't want to encourage customers to store a secret (the key) somewhere when managed identity is available.

                    key: azureSearchAdminKey,
                }
            }
        }
    ]
}
```

### Using Azure OpenAI client

> We will access Azure Open AI service through `openai` library. To authenticate our client, we will need to import `@azure/identity` to use Microsoft Entra ID token authentication. To learn more about the credential, refer to the README.md for the package [here]( https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/README.md).

Reviewer 2: ```suggestion
We'll access the Azure Open AI service through the `openai` library. To authenticate our client, we will need to import `@azure/identity` to use Microsoft Entra ID token authentication. To learn more about the credential, refer to the README.md for the package [here]( https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/README.md).
```

> We will access Azure Open AI service through `openai` library. To authenticate our client, we will need to import `@azure/identity` to use Microsoft Entra ID token authentication. To learn more about the credential, refer to the README.md for the package [here]( https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/README.md).

Reviewer 2: ```suggestion
We will access Azure Open AI service through `openai` library. To authenticate our client, we will need to import `@azure/identity` to use Microsoft Entra ID token authentication. To learn more about the library's credentials, refer to the README.md for the package [here]( https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/README.md).
```

> #### Authenticate using Microsoft Entra ID

Reviewer: This can be deleted.
```suggestion
```

```js

> const { AzureOpenAI } = require("openai");  

Reviewer 2: As a rule of thumb, use `import` instead of `require` in all Azure SDK-focused code samples for JS.

# Use Azure OpenAI On Your Data With openai-node Library

## Objective

Use openai-node Azure client to generate answers from your data using Azure On Your Data feature.

This tutorial uses the following services:

- Access to Azure OpenAI Service - you can apply for access [here](https://go.microsoft.com/fwlink/?linkid=2222006)
- Access to Azure Search Service 

## Time

You should expect to spend 5-10 minutes running this sample. 

## Before you begin

### Set Environment Variables

To start with let us create a config `.env` file with your project details. This file can be used in this sample or other samples to connect to your workspace. 

> ```ts

Reviewer: ```suggestion
```js
```

// Load the .env file if it exists

> import * as dotenv from "dotenv";
> dotenv.config();

Reviewer: ```suggestion
import 'dotenv/config'
```

Reviewer: Also, I just noticed that the language listed is JavaScript but the code is typed as TypeScript, I think we should make sure it is proper JS code.

// You will need to set these environment variables or edit the following values
// The endpoint you will use to access your Azure OpenAI instance
const endpoint = process.env["ENDPOINT"] || "<endpoint>";
// Your Azure OpenAI API key

> const azureApiKey = process.env["AZURE_API_KEY"] || "<api key>";

Reviewer: Let's get rid of API key and promote only AAD. Also, let's let the client load the env var automatically. We can point to the user to set the `AZURE_OPENAI_ENDPOINT` var in their .env file.

// Your Azure Cognitive Search endpoint, admin key, and index name
const azureSearchEndpoint = process.env["AZURE_SEARCH_ENDPOINT"] || "<search endpoint>";
const azureSearchAdminKey = process.env["AZURE_SEARCH_KEY"] || "<search key>";
const azureSearchIndexName = process.env["AZURE_SEARCH_INDEX"] || "<search index>";
```

### Connect To Your Data Sources
We need to connect to a data source to upload our data. We will set up your own designated data sources using Azure AI Search. 
```ts
const dataSources = {
    data_sources: [
        {
            type: "azure_search",
            parameters: {
                endpoint: azureSearchEndpoint,
                index_name: azureSearchIndexName,
                authentication: {
                    type: "api_key",
                    key: azureSearchAdminKey,
                }
            }
        }
    ]
}
```
### Using Azure OpenAI client
We will access Azure Open AI service through `openai-node` library. 

> #### Authenticate Using API Key

Reviewer: Should we delete this section now that we're pushing for Entra?

```ts
import { AzureOpenAI } from 'openai';

// Make sure to set both AZURE_OPENAI_ENDPOINT with the endpoint of your Azure resource and AZURE_OPENAI_API_KEY with the API key.

> // You can find both information in the Azure Portal.

Reviewer: Let's move the comments from code to the surrounding markdown.

const client = new AzureOpenAI({
    apiKey: azureApiKey,
    apiVersion,
    endpoint,
});
```

#### Authenticate Using Microsoft Entra ID

> ```ts

Reviewer: ```suggestion
```js
```

> import { AzureOpenAI } from 'openai';
> import { getBearerTokenProvider, DefaultAzureCredential } from "@azure/identity";

Reviewer: ```suggestion
const { AzureOpenAI } = require("openai");
const { getBearerTokenProvider, DefaultAzureCredential } = require("@azure/identity");
```

const credential = new DefaultAzureCredential();      
const scope = "https://cognitiveservices.azure.com/.default";

const client = new AzureOpenAI({
    azureADTokenProvider: getBearerTokenProvider(credential, scope),

>     apiVersion,

Reviewer: Where is this defined?

>     endpoint,

Reviewer: We can remove this now.

});
```

## Run AI Models On Your Data

> Now we can use Azure on your own data with Chat Completions. Providing our search endpoint, key, and index name in dataSources, any questions posed to the model will now be grounded in our own data. An additional property, context, will be provided in the response to show the data the model referenced to answer the question.

Reviewer: ```suggestion
Now we can use Azure on your own data with Chat Completions. Providing our search endpoint, key, and index name in the `data_sources` options, any questions posed to the model will now be grounded in our own data. An additional property, `context`, will be provided in the response to show the data the model referenced to answer the question.
```

> ```ts

Reviewer: ```suggestion
```js
```

async function main() {
    console.log('== Get chat completions on your own data ==');
    const result = await client.chat.completions.create({

>         model: "gpt-4",

Reviewer: I think we should emphasize that the value we pass to the model property is the deployment name. We can even have it in a variable named `deploymentName`.

        messages: [{ role: 'user', content: 'Write a poem to celebrate my birthday!' }],

>         ...dataSources as any

Reviewer: ```suggestion
        ...dataSources
```

    });

>     console.log("The content received:", result.choices[0]!.message?.content);

Reviewer: Is `!` needed?
```suggestion
    console.log("The content received:", result.choices[0].message?.content);
```
